Issue #12918 NPE in AuthenticationState.ServeAs.wrap

janbartel opened the pull request:
Fixes #12918

As the `ServletContextRequest` fields `_requestedSession` and `_managedSession` are `private`, they must be explicitly set on the newly created instance in `ServletContextHandler.newServletContextRequest` when the passed-in request is itself a `ServletContextRequest`.

janbartel committed:
Issue #12918 NPE in AuthenticationState.ServeAs.wrap

joakime commented:
@gregw @janbartel ping, are we good to get this into 12.0.20?

gregw commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextRequest.java:
```diff
@@ -190,6 +190,11 @@ public HttpURI getHttpURI()
             decodedPathInContext,
             matchedResource
         );
+        if (request instanceof ServletContextRequest originalServletContextRequest)
```
But this will not work if the request is further wrapped.   So you probably need something like:
```suggestion
        ServletContextRequest originalServletContextRequest = Request.as(request, ServletContextRequest.class);
        if (originalServletContextRequest != null)
```

The `Request.as` will unwrap wrappers looking for the type.... but now I'm worried that will see through a cross context dispatch.  Perhaps we need to modify as so that it does not see through cross context dispatch?

janbartel replied:
@gregw ok, so how about we check that a context boundary isn't crossed as we do the unwrap?  See commit https://github.com/jetty/jetty.project/pull/13041/commits/62852e56321260f65b8cd5d22a30148b3f222e23

janbartel committed:
Ensure context boundary not crossed in unwrap.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1033,6 +1033,27 @@ static <T> T as(Request request, Class<T> type)
+     * @param context the limiting context
+     * @return the request unwrapped back to the given type, or null if it cannot be
+     * unwrapped to that type or a context boundary is crossed.
+     */
```
Add a @see as(Request,Class<T>) and can you give some javadoc for that method  too?

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1033,6 +1033,27 @@ static <T> T as(Request request, Class<T> type)
+     */
+    static <T extends Request> T as(Request request, Class<T> type, Context context)
+    {
+        while (request != null)
+        {
+            if (type.isInstance(request))
+                return (T)request;
+            request = request instanceof Request.Wrapper wrapper ? wrapper.getWrapped() : null;
+            if (request == null || request.getContext() != context)
+                return null;
+        }
+        return null;
```
I've change my mind on the order of the test, as it will avoid the double null check:

```suggestion
        while (request != null)
        {
            if (request.getContext() != context)
                return null;
            if (type.isInstance(request))
                return (T)request;
            request = request instanceof Request.Wrapper wrapper ? wrapper.getWrapped() : null;
        }
        return null;
```

gregw reviewed (changes requested):
Can you look through any current callers of `as(...)` to see if there are others that should not cross a context boundary

janbartel committed:
Update jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

Co-authored-by: Greg Wilkins <gregw@webtide.com>

janbartel committed:
Add javadoc to existing Request.as method

joakime commented:
As there is only 1 day left in this month, this PR is being pushed to next month's 12.0.21 release.

janbartel commented:
> Can you look through any current callers of `as(...)` to see if there are others that should not cross a context boundary

Almost all of the usages of `Request.as` come from `ee10`, where we always wrap the inbound request on a cross context dispatch. So where the code is unwrapping to other `ee10` classes, then they will be guaranteed to be the first one on the stack.

There are 4 other usages of `Request.as` in core, 2 of them trivial. The other 2 are in `EagerFormHandler` and `GzipHandler`, both of which _should_ cross a cross context boundary.

So in short I think we're good where we are?

joakime commented:
If the request enters ee10, gets cross-context-dispatched to something in ee9, which then also cross-context-dispatches to some other ee10 context, is that a problem?

gregw commented:
IF we are looking for anything contexty (e.g. servlet API), then we should not cross context boundaries.
But if we are looking for anything that is not context related (e.g. gzip) then we can cross context boundaries.

gregw reviewed (changes requested):
Can you add a test to check the assertion that we are ok with these unwraps in CCD?

janbartel commented:
> If the request enters ee10, gets cross-context-dispatched to something in ee9, which then also cross-context-dispatches to some other ee10 context, is that a problem?

This should not be a problem. ee9 code does not use these `Request` methods, and ee10 wraps every time.

janbartel commented:
> Can you add a test to check the assertion that we are ok with these unwraps in CCD?

It's actually very difficult to write tests for this, as the use of `Request.as` is way down in jetty code. Any tests can only indirectly confirm that the correct unwrap was done, rather than test directly.

I propose that we formalize your statement: 

> IF we are looking for anything contexty (e.g. servlet API), then we should not cross context boundaries.
> But if we are looking for anything that is not context related (e.g. gzip) then we can cross context boundaries.

By changing all uses of `Request.as` to match: if accessing anything "contexty" then we use the 3 param arg version of the method to ensure we don't cross a context boundary, otherwise we just use the current 2 param method (eg for `GzipHandler` and `EagerFormHandler`).

janbartel committed:
Modify after discussion; add tests

janbartel commented:
@gregw I have implemented the changes we discussed, and added 2 unit tests: one for `Request.asInContext` and the other for `Response.asInContext`. Please re-review.
